File: backend/src/main/java/project/backend/domain/chat/chatmessage/api/ChatMessageController.java
package project.backend.domain.chat.chatmessage.api;

import java.security.Principal;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.messaging.handler.annotation.DestinationVariable;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import project.backend.domain.chat.chatmessage.app.ChatMessageService;
import project.backend.domain.chat.chatmessage.dto.ChatMessageEditRequest;
import project.backend.domain.chat.chatmessage.dto.ChatMessageRequest;
import project.backend.domain.chat.chatmessage.dto.ChatMessageResponse;
import project.backend.domain.chat.chatmessage.dto.ChatMessageSearchRequest;
import project.backend.domain.chat.chatmessage.dto.ChatMessageSearchResponse;
import project.backend.domain.imagefile.ImageFile;
import project.backend.domain.imagefile.ImageFileService;
import project.backend.domain.imagefile.ImageType;
import project.backend.global.config.security.dto.MemberDetails;

@RestController
@RequiredArgsConstructor
public class ChatMessageController {

	private final ChatMessageService chatMessageService;
	private final ImageFileService imageFileService;
	private final SimpMessagingTemplate messagingTemplate;

	@MessageMapping("/send-message/{roomId}") //클라이언트가 메세지를 보낼 경로
	public ChatMessageResponse sendMessage(@DestinationVariable Long roomId,
		@Payload ChatMessageRequest request, Principal principal) {
		ChatMessageResponse response = chatMessageService.save(roomId, request,
			principal.getName());

		messagingTemplate.convertAndSend("/topic/chat/" + roomId, response);

		return response;
	}

	@MessageMapping("/edit-message/{roomId}")
	public ChatMessageResponse editMessage(@DestinationVariable Long roomId, @Payload
	ChatMessageEditRequest request, Principal principal) {
		ChatMessageResponse response = chatMessageService.editMessage(roomId, request,
			principal.getName());

		messagingTemplate.convertAndSend("/topic/chat/" + roomId, response);

		return response;
	}

	@GetMapping("/chat/search/{roomId}")
	public Page<ChatMessageSearchResponse> searchMessages(
		@PathVariable("roomId") Long roomId,
		@RequestParam("keyword") String keyword,
		@RequestParam(defaultValue = "0") int page,
		@RequestParam(defaultValue = "10") int size
	) {
		ChatMessageSearchRequest request = ChatMessageSearchRequest.of(keyword, page, size);

		return chatMessageService.searchMessages(roomId, request);
	}

	@GetMapping("/{roomId}/messages")
	public List<ChatMessageResponse> getMessages(@PathVariable Long roomId,
		@AuthenticationPrincipal MemberDetails memberDetails) {
		return chatMessageService.getMessagesByRoomId(roomId, memberDetails.getId());
	}

	@MessageMapping("/delete-message/{roomId}")
	public ChatMessageResponse deleteMessage(@DestinationVariable Long roomId,
		@Payload Long messageId,
		Principal principal) {
		ChatMessageResponse response = chatMessageService.deleteMessage(roomId, messageId,
			principal.getName());

		messagingTemplate.convertAndSend("/topic/chat/" + roomId, response);

		return response;
	}

	@PostMapping("/send-image")
	public Long uploadImage(@RequestParam MultipartFile image) {
		ImageFile imageFile = imageFileService.saveImageFile(image, ImageType.CHAT_IMAGE);
		return imageFile.getImageId();
	}

}

File: backend/src/main/java/project/backend/domain/chat/chatmessage/api/ChatMessageController.kt
package project.backend.domain.chat.chatmessage.api

import ChatMessageResponse
import org.springframework.data.domain.Page
import org.springframework.messaging.handler.annotation.DestinationVariable
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import project.backend.auth.dto.MemberDetails
import project.backend.domain.chat.chatmessage.app.ChatMessageService
import project.backend.domain.chat.chatmessage.dto.*
import project.backend.domain.imagefile.ImageFileService
import java.security.Principal

@RestController
class ChatMessageController(
    private val chatMessageService: ChatMessageService,
    private val imageFileService: ImageFileService,
    private val messagingTemplate: SimpMessagingTemplate
) {

    @MessageMapping("/send-message/{roomId}")
    fun sendMessage(
        @DestinationVariable roomId: Long,
        @Payload request: ChatMessageRequest,
        principal: Principal
    ): ChatMessageResponse {
        val response = chatMessageService.save(roomId, request, principal.name)
        messagingTemplate.convertAndSend("/topic/chat/$roomId", response)
        return response
    }

    @MessageMapping("/edit-message/{roomId}")
    fun editMessage(
        @DestinationVariable roomId: Long,
        @Payload request: ChatMessageEditRequest,
        principal: Principal
    ): ChatMessageResponse {
        val response = chatMessageService.editMessage(roomId, request, principal.name)
        messagingTemplate.convertAndSend("/topic/chat/$roomId", response)
        return response
    }

    @GetMapping("/chat/search/{roomId}")
    fun searchMessages(
        @AuthenticationPrincipal memberDetails: MemberDetails,
        @PathVariable("roomId") roomId: Long,
        @RequestParam("keyword") keyword: String,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int
    ): Page<ChatMessageSearchResponse> {
        val request = ChatMessageSearchRequest.of(keyword, page, size)
        return chatMessageService.searchMessages(memberDetails.id, roomId, request)
    }

    @GetMapping("/{roomId}/messages")
    fun getMessages(
        @AuthenticationPrincipal memberDetails: MemberDetails,
        @PathVariable roomId: Long,
        @RequestParam(required = false) cursor: Long?,
        @RequestParam(defaultValue = "30") size: Int
    ): ChatScrollResponse {
        return chatMessageService.getMessagesByRoomId(memberDetails.id, roomId, cursor, size)
    }

    @MessageMapping("/delete-message/{roomId}")
    fun deleteMessage(
        @DestinationVariable roomId: Long,
        @Payload messageId: Long,
        principal: Principal
    ): ChatMessageResponse {
        val response = chatMessageService.deleteMessage(roomId, messageId, principal.name)
        messagingTemplate.convertAndSend("/topic/chat/$roomId", response)
        return response
    }

    @PostMapping("/send-image")
    fun uploadImage(@RequestParam image: MultipartFile): Long? {
        val imageFile = imageFileService.saveChatImage(image)
        return imageFile.imageId
    }
}

